### PR TITLE
[FLINK-12939][docs-zh] Translate "Apache Kafka Connector" page into Chinese

### PR DESCRIPTION
**Pull request by yangjf2019**
## What is the purpose of the change

*Translate "Apache Kafka Connector" page into Chinese*

## Brief change log

  - *I modified the kafka.zh.md file .*

## Verifying this change

*I have verified the changes by starting the build script in preview mode.*

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): ( no )
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: ( no )
  - The serializers: ( no )
  - The runtime per-record code paths (performance sensitive): ( no )
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: ( no )
  - The S3 file system connector: ( no )

## Documentation

  - Does this pull request introduce a new feature? (  no )
  - If yes, how is the feature documented? ( not documented )

Hi,  @wuchong @TisonKun ,I replace a new account, so open a new PR, but the content of file is not change, here is the  old [PR](https://github.com/apache/flink/pull/9415) . Please Take A Look .

### CODE REVIEW

File: docs/dev/connectors/kafka.zh.md [hunk 1]
```diff
@@ -448,61 +383,36 @@ val stream = env.addSource(myConsumer)
...
+
+配置 offset 提交行为的方法是否相同，取决于是否为 job 启用了 checkpointing。
+
+ - *禁用 Checkpointing：* 如果禁用了 checkpointing，则 Flink Kafka Consumer 依赖于内部使用的 Kafka client 自动定期 offset 提交功能。
+ 因此，要禁用或启用 offset 的提交，只需将 `enable.auto.commit`（或 Kafka 0.8 的 `auto.commit.enable`）或者 `auto.commit.interval.ms` 的Key 值设置为提供的 `Properties` 配置中的适当值。
+
+ - *启用 Checkpointing：* 如果启用了 checkpointing，那么当 checkpointing 完成时，Flink Kafka Consumer 将提交的 offset 存储在 checkpoint 状态中。
+ 这确保 Kafka broker 中提交的 offset 与 checkpoint 状态中的 offset 一致。
+ 用户可以通过调用 consumer 上的 `setCommitOffsetsOnCheckpoints(boolean)` 方法来禁用或启用 offset 的提交(默认情况下，这个值是 true )。
+ 注意，在这个场景中，完全忽略 `Properties` 中的自动定期 offset 提交设置。
```
Comment (wuchong) [inline docs/dev/connectors/kafka.zh.md]:
```suggestion
 注意，在这个场景中，`Properties` 中的自动定期 offset 提交配置会被完全忽略。
```

File: docs/dev/connectors/kafka.zh.md [hunk 2]
```diff
@@ -448,61 +383,36 @@ val stream = env.addSource(myConsumer)
...
+
+### Kafka Consumer 和 Timestamp extractor 或者 Watermark Emission
+
+在许多场景中，记录的时间戳(显式或隐式)嵌入到记录本身中。此外，用户可能希望定期或以不规则的方式 Watermark Emission，例如基于 Kafka 流中包含当前事件时间水位线的特殊记录。对于这些情况，Flink Kafka Consumer 允许指定 `AssignerWithPeriodicWatermarks` 或 `AssignerWithPunctuatedWatermarks`。
```
Comment (wuchong) [inline docs/dev/connectors/kafka.zh.md]:
```suggestion
在许多场景中，记录的时间戳是（显式或隐式地）嵌入到记录自身中的。此外，用户可能希望定期或以不规则的方式发送 watermark，例如基于 Kafka 流中包含当前事件时间的 watermark 的特殊记录。对于这些情况，Flink Kafka Consumer 允许指定 `AssignerWithPeriodicWatermarks` 或 `AssignerWithPunctuatedWatermarks`。
```

File: docs/dev/connectors/kafka.zh.md [hunk 3]
```diff
@@ -448,61 +383,36 @@ val stream = env.addSource(myConsumer)
...
+
+你可以按照[此处]({{ site.baseurl }}/dev/event_timestamps_watermarks.html)的说明指定自定义 timestamp extractor 或者 Watermark Emission，或者使用 [predefined ones]({{ site.baseurl }}/dev/event_timestamp_extractors.html)。你也可以通过以下方式将其传递给你的 consumer：
```
Comment (wuchong) [inline docs/dev/connectors/kafka.zh.md]:
```suggestion
你可以按照[此处]({{ site.baseurl }}/zh/dev/event_timestamps_watermarks.html)的说明指定自定义时间戳抽取器或 watermark 发送器，或者使用 [内置的]({{ site.baseurl }}/dev/event_timestamp_extractors.html)。你可以通过以下方式将其传递给你的 consumer：
```

File: docs/dev/connectors/kafka.zh.md [hunk 4]
```diff
@@ -577,171 +474,101 @@ stream.addSink(myProducer);
...
-To do this, provide a `null` custom partitioner when instantiating the producer. It is important
-to provide `null` as the custom partitioner; as explained above, if a custom partitioner is not specified
-the `FlinkFixedPartitioner` is used instead.
-
-### Kafka Producers and Fault Tolerance
+上面的例子演示了创建 Flink Kafka Producer 来将流消息写入单个 Kafka 目标 topic 的基本用法。
+对于更高级的用法，这还有其他构造函数变体允许提供以下内容：
+
+ * *提供自定义属性*：producer 允许为内部 `KafkaProducer` 提供自定义属性配置。有关如何配置 Kafka Producer 的详细信息，请参阅  [Apache Kafka 文档](https://kafka.apache.org/documentation.html)。
+ * *自定义分区器*：要将消息分配给特定的分区，可以向构造函数提供一个 `FlinkKafkaPartitioner` 的实现。这个分区器将被流中的每条记录调用，以确定消息应该发送到目标 topic 的哪个具体分区里。有关详细信息，请参阅 [Kafka Producer 分区方案](#kafka-producer-partitioning-scheme)。
```
Comment (wuchong) [inline docs/dev/connectors/kafka.zh.md]:
```suggestion
 * *自定义分区器*：要将消息分配给特定的分区，可以向构造函数提供一个 `FlinkKafkaPartitioner` 的实现。这个分区器将被流中的每条记录调用，以确定消息应该发送到目标 topic 的哪个具体分区里。有关详细信息，请参阅 [Kafka Producer 分区方案](#kafka-producer-分区方案)。
```

File: docs/dev/connectors/kafka.zh.md [hunk 5]
```diff
@@ -577,171 +474,101 @@ stream.addSink(myProducer);
...
+ * `Semantic.EXACTLY_ONCE`：使用 Kafka 事务提供精准一次的语义。无论何时，在使用事务写入 Kafka 时，都要记得为所有消费 Kafka 消息的应用程序设置所需的 `isolation.level`（ `read_committed` 或 `read_uncommitted`  - 后者是默认值）。
+
+##### 注意事项
+
+`Semantic.EXACTLY_ONCE` 模式依赖于事务提交的能力。事务提交发生于触发 checkpoint 之前，以及从 checkpoint 恢复之后。如果从 Flink 应用程序崩溃到完全重启的时间超过了 Kafka 的事务超时时间，那么将会有数据丢失（Kafka 会自动丢弃超出超时时间的事务）。考虑到这一点，请根据预期的宕机时间来合理地配置事务超时时间。
+
+默认情况下，Kafka broker 将 `transaction.max.timeout.ms` 设置为 15 分钟。此属性不允许为大于其值的 producer 设置事务超时时间。
+默认情况下，`FlinkKafkaProducer011` 将 producer config 中的 `transaction.timeout.ms` 属性设置为 1 小时，因此在使用 `Semantic.EXACTLY_ONCE` 模式之前应该增加 `transaction.max.timeout.ms` 的值。
+
+在 `KafkaConsumer` 的 `read_committed` 模式中，任何未完成的事务（既不中止也不完成）将阻止来自给定 Kafka  topic 的所有读取超过任何未完成的事务。
```
Comment (wuchong) [inline docs/dev/connectors/kafka.zh.md]:
```suggestion
在 `KafkaConsumer` 的 `read_committed` 模式中，任何未结束（既未中止也未完成）的事务将阻塞来自给定 Kafka topic 的未结束事务之后的所有读取数据。
```

File: docs/dev/connectors/kafka.zh.md [hunk 6]
```diff
@@ -577,171 +474,101 @@ stream.addSink(myProducer);
...
+1. 用户启动了 `transaction1` 并使用它写了一些记录
+2. 用户启动了 `transaction2` 并使用它编写了一些其他记录
+3. 用户提交了 `transaction2`
+
+即使 `transaction2` 中的记录已提交，在提交或中止 `transaction1` 之前，消费者也不会看到这些记录。这有 2 层含义：
+
+ * 首先，在 Flink 应用程序的正常工作期间，用户可以预料 Kafka 主题中生成的记录的可见性会延迟，相当于已完成 checkpoint 之间的平均时间。
+ * 其次，在 Flink 应用程序失败的情况下，此应用程序正在写入的供消费者读取的主题将被阻塞，直到应用程序重新启动或配置的事务超时时间过去后，才恢复正常。此标注仅适用于有多个 agent 或者应用程序写入同一 Kafka 主题的情况。
+
+**注意**：`Semantic.EXACTLY_ONCE` 模式为每个 `FlinkKafkaProducer011` 实例使用固定大小的 KafkaProducer 池。每个 checkpoint 使用其中一个 producer。如果并发检查点的数量超过池的大小，`FlinkKafkaProducer011` 将抛出异常，并导致整个应用程序失败。请合理地配置最大池大小和最大并发检查点数量。
```
Comment (wuchong) [inline docs/dev/connectors/kafka.zh.md]:
```suggestion
**注意**：`Semantic.EXACTLY_ONCE` 模式为每个 `FlinkKafkaProducer011` 实例使用固定大小的 KafkaProducer 池。每个 checkpoint 使用其中一个 producer。如果并发 checkpoint 的数量超过池的大小，`FlinkKafkaProducer011` 将抛出异常，并导致整个应用程序失败。请合理地配置最大池大小和最大并发 checkpoint 数量。
```
专有名词 翻译或不翻译，请保持上下文中的一致。

File: docs/dev/connectors/kafka.zh.md [hunk 7]
```diff
@@ -577,171 +474,101 @@ stream.addSink(myProducer);
...
+
+**注意**：`Semantic.EXACTLY_ONCE` 采取一切可能的措施，不留下任何会阻止消费者读取 Kafka 主题消息的延迟事务，那么这是有必要的。但是，如果在第一个 checkpoint 之前 Flink 应用程序失败，则在重新启动此类应用程序后，系统中没有关于先前池大小的信息。因此，在第一个 checkpoint 完成之前缩小 Flink 应用程序是不安全的，因为大于了 `FlinkKafkaProducer011.SAFE_SCALE_DOWN_FACTOR` 的安全系数。
```
Comment (wuchong) [inline docs/dev/connectors/kafka.zh.md]: 这段话很费解，请再改进一下。

File: docs/dev/connectors/kafka.zh.md [hunk 8]
```diff
@@ -448,61 +383,36 @@ val stream = env.addSource(myConsumer)
...
+
+ - *禁用 Checkpointing：* 如果禁用了 checkpointing，则 Flink Kafka Consumer 依赖于内部使用的 Kafka client 自动定期 offset 提交功能。
+ 因此，要禁用或启用 offset 的提交，只需将 `enable.auto.commit`（或 Kafka 0.8 的 `auto.commit.enable`）或者 `auto.commit.interval.ms` 的Key 值设置为提供的 `Properties` 配置中的适当值。
+
+ - *启用 Checkpointing：* 如果启用了 checkpointing，那么当 checkpointing 完成时，Flink Kafka Consumer 将提交的 offset 存储在 checkpoint 状态中。
+ 这确保 Kafka broker 中提交的 offset 与 checkpoint 状态中的 offset 一致。
+ 用户可以通过调用 consumer 上的 `setCommitOffsetsOnCheckpoints(boolean)` 方法来禁用或启用 offset 的提交(默认情况下，这个值是 true )。
+ 注意，在这个场景中，完全忽略 `Properties` 中的自动定期 offset 提交设置。
+
+### Kafka Consumer 和 Timestamp extractor 或者 Watermark Emission
```
Comment (wuchong) [inline docs/dev/connectors/kafka.zh.md]:
```suggestion
### Kafka Consumer 和时间戳抽取以及 watermark 发送
```

File: docs/dev/connectors/kafka.zh.md [hunk 9]
```diff
@@ -577,171 +474,101 @@ stream.addSink(myProducer);
...
+
+**注意**：`Semantic.EXACTLY_ONCE` 采取一切可能的措施，不留下任何会阻止消费者读取 Kafka 主题消息的延迟事务，那么这是有必要的。但是，如果在第一个 checkpoint 之前 Flink 应用程序失败，则在重新启动此类应用程序后，系统中没有关于先前池大小的信息。因此，在第一个 checkpoint 完成之前缩小 Flink 应用程序是不安全的，因为大于了 `FlinkKafkaProducer011.SAFE_SCALE_DOWN_FACTOR` 的安全系数。
+
+## 在 Kafka 0.10.x 中使用 Kafka 时间戳和 Flink 事件时间
+
+自 Apache Kafka 0.10+ 以来，Kafka 的消息可以携带[时间戳](https://cwiki.apache.org/confluence/display/KAFKA/KIP-32+-+Add+timestamps+to+Kafka+message)，指示事件发生的时间（请参阅 Apache Flink 中的"事件时间"(../event_time.html)）或消息写入 Kafka broker 的时间。
+
+如果 Flink 中的时间特性设置为 `TimeCharacteristic.EventTime`（ `StreamExecutionEnvironment.setStreamTimeCharacteristic（TimeCharacteristic.EventTime）`），则 `FlinkKafkaConsumer010` 将发出附加时间戳的记录。
+
+Kafka consumer 不会发出水位线。为了发出水位线，可采用 `assignTimestampsAndWatermarks` 方法的 "Kafka Consumer and Timestamp Extraction / Watermark Emission" 中描述的相同机制的机制。
```
Comment (wuchong) [inline docs/dev/connectors/kafka.zh.md]:
```suggestion
Kafka consumer 不会发出 watermark。为了发出 watermark，可采用上文 “Kafka Consumer 和时间戳抽取以及 watermark 发送” 章节中描述的 `assignTimestampsAndWatermarks` 方法。
```

File: docs/dev/connectors/kafka.zh.md [hunk 10]
```diff
@@ -750,7 +577,7 @@ public long extractTimestamp(Long element, long previousElementTimestamp) {
 
 
 
-The `FlinkKafkaProducer010` only emits the record timestamp, if `setWriteTimestampToKafka(true)` is set.
+如果设置了 `setWriteTimestampToKafka(true)`，则 `FlinkKafkaProducer010` 仅发出记录时间戳。
```
Comment (wuchong) [inline docs/dev/connectors/kafka.zh.md]:
```suggestion
只有设置了 `setWriteTimestampToKafka(true)`，则 `FlinkKafkaProducer010` 才会发出记录时间戳。
```

File: docs/dev/connectors/kafka.zh.md [hunk 11]
```diff
@@ -759,85 +586,50 @@ config.setWriteTimestampToKafka(true);
...
 
 
-## Kafka Connector metrics
-
-Flink's Kafka connectors provide some metrics through Flink's [metrics system]({{ site.baseurl }}/monitoring/metrics.html) to analyze
-the behavior of the connector.
-The producers export Kafka's internal metrics through Flink's metric system for all supported versions. The consumers export 
-all metrics starting from Kafka version 0.9. The Kafka documentation lists all exported metrics 
-in its [documentation](http://kafka.apache.org/documentation/#selector_monitoring).
+## Kafka Connector 指标
```
Comment (wuchong) [inline docs/dev/connectors/kafka.zh.md]:
```suggestion
## Kafka 连接器指标
```

File: docs/dev/connectors/kafka.zh.md [hunk 12]
```diff
@@ -759,85 +586,50 @@ config.setWriteTimestampToKafka(true);
...
 
-In addition to these metrics, all consumers expose the `current-offsets` and `committed-offsets` for each topic partition.
-The `current-offsets` refers to the current offset in the partition. This refers to the offset of the last element that
-we retrieved and emitted successfully. The `committed-offsets` is the last committed offset.
+Flink 的 Kafka 连接器通过 Flink 的 [metric 系统]({{ site.baseurl }}/monitoring/metrics.html) 提供一些指标来分析 Kafka Connector 的状况。producer 通过 Flink 的 metrics 系统为所有支持的版本导出 Kafka 的内部指标。consumer 从 Kafka 0.9 版本开始导出所有指标。Kafka 文档在其[文档](http://kafka.apache.org/documentation/#selector_monitoring)中列出了所有导出的指标。
```
Comment (wuchong) [inline docs/dev/connectors/kafka.zh.md]:
```suggestion
Flink 的 Kafka 连接器通过 Flink 的 [metric 系统]({{ site.baseurl }}/zh/monitoring/metrics.html) 提供一些指标来分析 Kafka Connector 的状况。Producer 通过 Flink 的 metrics 系统为所有支持的版本导出 Kafka 的内部指标。Consumer 从 Kafka 0.9 版本开始导出所有指标。Kafka 文档在其[文档](http://kafka.apache.org/documentation/#selector_monitoring)中列出了所有导出的指标。
```

File: docs/dev/connectors/kafka.zh.md [hunk 13]
```diff
@@ -759,85 +586,50 @@ config.setWriteTimestampToKafka(true);
...
 
-The Kafka Consumers in Flink commit the offsets back to Zookeeper (Kafka 0.8) or the Kafka brokers (Kafka 0.9+). If checkpointing
-is disabled, offsets are committed periodically.
-With checkpointing, the commit happens once all operators in the streaming topology have confirmed that they've created a checkpoint of their state. 
-This provides users with at-least-once semantics for the offsets committed to Zookeeper or the broker. For offsets checkpointed to Flink, the system 
-provides exactly once guarantees.
+除了这些指标之外，所有 consumer 都会公开每个主题分区的 `current-offsets` 和 `committed-offsets`。`current-offsets` 是指分区中的当前偏移量。指的是我们成功检索和发出的最后一个元素的偏移量。`committed-offsets` 是最后提交的偏移量。这为用户提供了 at-least-once 语义，用于提交给 Zookeeper 或 broker 的偏移量。对于 Flink 的偏移检查点，系统提供精准一次语义。
```
Comment (wuchong) [inline docs/dev/connectors/kafka.zh.md]:
```suggestion
除了这些指标之外，所有 consumer 都暴露了每个主题分区的 `current-offsets` 和 `committed-offsets`。`current-offsets` 是指分区中的当前偏移量。指的是我们成功检索和发出的最后一个元素的偏移量。`committed-offsets` 是最后提交的偏移量。这为用户提供了 at-least-once 语义，用于提交给 Zookeeper 或 broker 的偏移量。对于 Flink 的偏移检查点，系统提供精准一次语义。
```

File: docs/dev/connectors/kafka.zh.md [hunk 14]
```diff
@@ -759,85 +586,50 @@ config.setWriteTimestampToKafka(true);
...
-- Set `sasl.kerberos.service.name` to `kafka` (default `kafka`): The value for this should match the `sasl.kerberos.service.name` used for Kafka broker configurations.
-A mismatch in service name between client and server configuration will cause the authentication to fail.
+- 将 `security.protocol` 设置为 `SASL_PLAINTEXT`（默认为 `NONE`）：用于与 Kafka broker 进行通信的协议。使用独立 Flink 部署时，也可以使用 `SASL_SSL`;请在[此处](https://kafka.apache.org/documentation/#security_configclients)查看如何为 SSL 配置 Kafka 客户端。
+- 将 `sasl.kerberos.service.name` 设置为 `kafka`（默认为 `kafka`）：此值应与用于 Kafka broker 配置的 `sasl.kerberos.service.name` 相匹配。客户端和服务器配置之间的服务名称不匹配将导致身份验证失败。
 
-For more information on Flink configuration for Kerberos security, please see [here]({{ site.baseurl}}/ops/config.html).
-You can also find [here]({{ site.baseurl}}/ops/security-kerberos.html) further details on how Flink internally setups Kerberos-based security.
+有关 Kerberos 安全性 Flink 配置的更多信息，请参见[这里]({{ site.baseurl}}/ops/config.html)。你也可以在[这里]({{ site.baseurl}}/ops/security-kerberos.html)进一步了解 Flink 如何在内部设置基于 kerberos 的安全性。
```
Comment (wuchong) [inline docs/dev/connectors/kafka.zh.md]:
```suggestion
有关 Kerberos 安全性 Flink 配置的更多信息，请参见[这里]({{ site.baseurl}}/zh/ops/config.html)。你也可以在[这里]({{ site.baseurl}}/zh/ops/security-kerberos.html)进一步了解 Flink 如何在内部设置基于 kerberos 的安全性。
```

File: docs/dev/connectors/kafka.zh.md [hunk 15]
```diff
@@ -759,85 +586,50 @@ config.setWriteTimestampToKafka(true);
...
 
 ## Troubleshooting
```
Comment (wuchong) [inline docs/dev/connectors/kafka.zh.md]:
```suggestion
## 问题排查
```

**Comment by yangjf2019** [comment]
Hi, @wuchong Thanks for your review, and left some comment in your free time.I will check it as soon as my possible.

File: docs/dev/connectors/kafka.zh.md [hunk 16]
```diff
@@ -26,25 +26,20 @@ under the License.
...
 
-Flink provides special Kafka Connectors for reading and writing data from/to Kafka topics.
-The Flink Kafka Consumer integrates with Flink's checkpointing mechanism to provide
-exactly-once processing semantics. To achieve that, Flink does not purely rely on Kafka's consumer group
-offset tracking, but tracks and checkpoints these offsets internally as well.
+Flink 提供了专门的 Kafka 连接器，向 Kafka topic 中读取或者写入数据。Flink Kafka Consumer 集成了 Flink 的 Checkpoint 机制，可提供 exactly-once 的处理语义。为此，Flink 并不完全依赖于跟踪 Kafka 消费组的偏移量，而是在内部跟踪和检查偏移量。
 
-Please pick a package (maven artifact id) and class name for your use-case and environment.
-For most users, the `FlinkKafkaConsumer08` (part of `flink-connector-kafka`) is appropriate.
+为你的用例和环境选择相应的包（maven artifact id）和类名。对于大多数用户来说，使用 `FlinkKafkaConsumer08`（ `flink-connector-kafka` 的一部分）是比较合适的。
```
[the remark this answers is not in the document]
Comment (yangjf2019) [reply docs/dev/connectors/kafka.zh.md]: Thanks ,I resolved them.

File: docs/dev/connectors/kafka.zh.md (hunk 6 above)
Comment (yangjf2019) [reply docs/dev/connectors/kafka.zh.md]: Thanks ,I resolved them.

File: docs/dev/connectors/kafka.zh.md (hunk 7 above)
Comment (yangjf2019) [reply docs/dev/connectors/kafka.zh.md]: 谢谢 @wuchong  抽空帮 review，这段话，我觉得如下翻译，可能合理一些，但是有一句表达确实费解（Thus it is unsafe to scale down Flink
application before first checkpoint completes  ----》减少 Flink 应用程序），麻烦再讨论下，感谢🙏

**注意**：`Semantic.EXACTLY_ONCE` 采取所有可能的措施，以免留下任何挥之不去的事务，否则将有必要阻止更多的消费者从 Kafka topic 中订阅内容。但是，如果 Flink 应用程序在第一次 checkpoint 之前就失败了，那么在重新启动此类应用程序后，系统中就不会有与先前池大小相关的信息。因此，在第一次 checkpoint 完成前按比例减少 Flink 应用程序是不安全的，因为超过了安全系数 `FlinkKafkaProducer011.SAFE_SCALE_DOWN_FACTOR` 的值。

Comment (wuchong) [reply docs/dev/connectors/kafka.zh.md]: I will improve this a bit and merge.

**Comment by yangjf2019** [comment]
@wuchong  Thanks for your helping :)

**Pull request closed without merging**